File: sdk/python/symbolchain/symbol/Metadata.py
import hashlib


def metadata_generate_key(seed):
	"""Generates a metadata key from a string."""

	hash_result = hashlib.sha3_256(seed.encode('utf8')).digest()

	key_bytes = bytearray(hash_result[:8])
	key_bytes[7] |= 0x80  # set high bit to match SDK V2 implementation

	return int.from_bytes(key_bytes, 'little')


def metadata_update_value(old_value, new_value):
	"""Creates a metadata payload for updating old value to new value."""

	if not old_value:
		return new_value

	shorter_length = min(len(old_value), len(new_value))
	longer_length = max(len(old_value), len(new_value))
	is_new_value_shorter = len(old_value) > len(new_value)

	result = bytearray()
	for i in range(shorter_length):
		result.append(old_value[i] ^ new_value[i])

	for i in range(shorter_length, longer_length):
		result.append((old_value if is_new_value_shorter else new_value)[i])

	return bytes(result)

File: sdk/javascript/src/symbol/metadata.js
import { sha3_256 } from '@noble/hashes/sha3.js';
import { utf8ToBytes } from '@noble/hashes/utils.js';

/**
 * Generates a metadata key from a string.
 * @param {string} seed Metadata key seed.
 * @returns {bigint} Metadata key.
 */
const metadataGenerateKey = seed => {
	const hashResult = sha3_256(utf8ToBytes(seed));

	const keyBytes = hashResult.subarray(0, 8);
	keyBytes[7] |= 0x80; // set high bit to match SDK V2 implementation

	const keys = new BigUint64Array(keyBytes.buffer);
	return keys[0];
};

/**
 * Creates a metadata payload for updating old value to new value.
 * @param {Uint8Array|undefined} oldValue Old metadata value.
 * @param {Uint8Array} newValue New metadata value.
 * @returns {Uint8Array} Metadata payload for updating old value to new value.
 */
const metadataUpdateValue = (oldValue, newValue) => {
	if (!oldValue)
		return newValue;

	const shorterLength = Math.min(oldValue.length, newValue.length);
	const longerLength = Math.max(oldValue.length, newValue.length);
	const isNewValueShorter = oldValue.length > newValue.length;

	const result = new Uint8Array(longerLength);

	let i = 0;
	for (i = 0; i < shorterLength; ++i)
		result[i] = oldValue[i] ^ newValue[i];

	for (; i < longerLength; ++i)
		result[i] = (isNewValueShorter ? oldValue : newValue)[i];

	return result;
};

export { metadataGenerateKey, metadataUpdateValue };
